Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp

```diff
@@ -0,0 +1,810 @@
+   {
+      delete NV_MFAB(v);
+      NV_MFAB(v) = NULL;
+   }
+   free(v->content); v->content = NULL;
+   free(v->ops); v->ops = NULL;
+   free(v); v = NULL;
+
+   return;
+}
+
+void N_VSpace_MultiFab(N_Vector v, sunindextype *lrw, sunindextype *liw)
+{
+   *lrw = NV_LENGTH_M(v);
+   *liw = 1;
+
+   return;
+}
+
+void N_VLinearSum_MultiFab(realtype a, N_Vector x, realtype b, N_Vector y,
+                           N_Vector z)
+{
+   amrex::MultiFab *mf_x = NV_MFAB(x);
+   amrex::MultiFab *mf_y = NV_MFAB(y);
+   amrex::MultiFab *mf_z = NV_MFAB(z);
+
+   sunindextype ncomp = mf_x->nComp();
+   //sunindextype nghost = mf_x->nGrow();
+   sunindextype nghost = 0;  // do not include ghost cells
+
+   amrex::MultiFab::LinComb(*mf_z, a, *mf_x, 0, b, *mf_y, 0, 0, ncomp, nghost);
+}
+
+void N_VConst_MultiFab(realtype c, N_Vector z)
+{
+   sunindextype i, N;
+   amrex::MultiFab *mf_z = NV_MFAB(z);
+   *mf_z = c;
+}
+
+void N_VProd_MultiFab(N_Vector x, N_Vector y, N_Vector z)
+{
+   amrex::MultiFab *mf_x = NV_MFAB(x);
+   amrex::MultiFab *mf_y = NV_MFAB(y);
+   amrex::MultiFab *mf_z = NV_MFAB(z);
+
+   sunindextype ncomp = mf_x->nComp();
+   //sunindextype nghost = mf_x->nGrow();
+   sunindextype nghost = 0;  // do not include ghost cells
+
+   amrex::MultiFab::Copy(*mf_z, *mf_x, 0, 0, ncomp, nghost);
+   amrex::MultiFab::Multiply(*mf_z, *mf_y, 0, 0, ncomp, nghost);
+}
+
+void N_VDiv_MultiFab(N_Vector x, N_Vector y, N_Vector z)
+{
+   amrex::MultiFab *mf_x = NV_MFAB(x);
+   amrex::MultiFab *mf_y = NV_MFAB(y);
+   amrex::MultiFab *mf_z = NV_MFAB(z);
+
+   sunindextype ncomp = mf_x->nComp();
+   //sunindextype nghost = mf_x->nGrow();
+   sunindextype nghost = 0;  // do not include ghost cells
+
+   amrex::MultiFab::Copy(*mf_z, *mf_x, 0, 0, ncomp, nghost);
+   amrex::MultiFab::Divide(*mf_z, *mf_y, 0, 0, ncomp, nghost);
+}
+
+void N_VScale_MultiFab(realtype c, N_Vector x, N_Vector z)
+{
+   amrex::MultiFab *mf_x = NV_MFAB(x);
+   amrex::MultiFab *mf_z = NV_MFAB(z);
+
+   sunindextype ncomp = mf_x->nComp();
+   //sunindextype nghost = mf_x->nGrow();
+   sunindextype nghost = 0;  // do not include ghost cells
+
+   amrex::MultiFab::Copy(*mf_z, *mf_x, 0, 0, ncomp, nghost);
+   mf_z->mult(c, 0, ncomp, nghost);
+}
+
+void N_VAbs_MultiFab(N_Vector x, N_Vector z)
+{
+   using namespace amrex;
+
+   MultiFab *mf_x = NV_MFAB(x);
+   MultiFab *mf_z = NV_MFAB(z);
+   sunindextype ncomp = mf_x->nComp();
+
+   // ghost cells not included
+   for (MFIter mfi(*mf_x); mfi.isValid(); ++mfi)
+   {
+      const amrex::Box& bx = mfi.validbox();
+      Array4<Real> const& x_fab = mf_x->array(mfi);
+      Array4<Real> const& z_fab = mf_z->array(mfi);
+      const auto lo = lbound(bx);
+      const auto hi = ubound(bx);
+
+      for (int c = 0; c < ncomp; ++c) {
+         for (int k = lo.z; k <= hi.z; ++k) {
+            for (int j = lo.y; j <= hi.y; ++j) {
+               for (int i = lo.x; i <= hi.x; ++i) {
+                  z_fab(i,j,k,c) = SUNRabs(x_fab(i,j,k,c));
```

Comment:
Why do we explicitly compute this on the CPU?  It seems that we should do it on GPU?  We should also update other functions for GPU.

Reply:
It looks like there's a FabArray function that should be used instead for the Abs routine:
https://github.com/AMReX-Codes/amrex/blob/development/Src/Base/AMReX_FabArrayUtility.H#L1288-L1294

For things like the weighted root mean square norm, I was looking at copying the GPU strategy used in `norminfmask` by changing the norm to multiply the weight instead of doing the max, and then having the squares and sqrt and dividing at the end by number of elements.
https://github.com/AMReX-Codes/amrex/blob/development/Src/Base/AMReX_BaseFab.H#L2271-L2310

For the more specialized routines, does it make a substantial difference if I use ParallelFor vs converting the entire MFIter loop to use ParallelForMF (given these should be large box sizes)?

Reply:
Changed cpu-only routines to use `ParallelFor`, `amrex::ReduceMin`, `amrex::ParReduce` etc.